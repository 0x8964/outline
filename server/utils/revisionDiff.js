// @flow
import fs from "fs";
import path from "path";
import { findLastIndex } from "lodash";
import diff from "node-htmldiff";
import { renderToHtml } from "rich-markdown-editor";

export default function revisionDiff(
  before: string,
  after: string,
  fullDiff: boolean = false,
  buffer: number = 1
) {
  // The basic idea here is to first render the Markdown to HTML, then diff the
  // HTML - both sides will have valid HTML so we should have a valid diff as well

  const beforeHtml = renderToHtml(before);
  const afterHtml = renderToHtml(after);
  const diffHtml = diff(beforeHtml, afterHtml);

  if (fullDiff) {
    return diffHtml;
  }

  if (before === after) {
    return "";
  }

  // split diff into individual lines and find the first and last changed tags
  // so we can chop around those lines rather than return the entire document.
  let lines = diffHtml.split("\n");

  const firstChangedLineIndex = findLastIndex(
    lines,
    (value) => value.includes("<ins>") || value.includes("<del>")
  );
  const lastChangedLineIndex = findLastIndex(
    lines,
    (value) => value.includes("</ins>") || value.includes("</del>")
  );

  return lines
    .slice(
      Math.max(0, firstChangedLineIndex - buffer),
      Math.min(lines.length, lastChangedLineIndex + buffer)
    )
    .join("\n");
}

export async function revisionDiffExample() {
  let before = await fs.promises.readFile(
    path.resolve(__dirname, "..", "test", "fixtures", "complex.md"),
    "utf8"
  );

  let after = await fs.promises.readFile(
    path.resolve(__dirname, "..", "test", "fixtures", "complexModified.md"),
    "utf8"
  );

  return revisionDiff(before, after);
}
