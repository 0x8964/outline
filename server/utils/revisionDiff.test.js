// @flow
import fs from "fs";
import path from "path";
import revisionDiff from "./revisionDiff";

it("should diff a new revision of a document", async () => {
  let fileValue1 = await fs.promises.readFile(
    path.resolve(__dirname, "..", "test", "fixtures", "complex.md"),
    "utf8"
  );

  let fileValue2 = await fs.promises.readFile(
    path.resolve(__dirname, "..", "test", "fixtures", "complexModified.md"),
    "utf8"
  );

  // expose API in lib, use direct path for now
  // !! == checklist item all need classes to distinguish

  const diff = revisionDiff(fileValue1, fileValue2);
  expect(diff).toMatchSnapshot();

  // TODO hand evaluate the veracity fo the snapshot
});

it("should return empty string when the new revision has the same content", () => {});
