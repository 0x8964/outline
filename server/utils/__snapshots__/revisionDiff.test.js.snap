// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should diff a new revision of a document 1`] = `
"<h1>Heading 1</h1>
<h2>Heading 2</h2>
<p>This is a test paragraph</p>
<p>This is a second test paragraph. This is a second sentence.</p>
<p>This is a another test paragraph. This is a another sentence.</p>
<ul>
<li>list item 1</li>
<li data-diff-node=\\"del\\" data-operation-index=\\"1\\"><del data-operation-index=\\"1\\">list item 2</del></li></ul>
<pre><code>this is a codeblock
</code></pre><p data-diff-node=\\"ins\\" data-operation-index=\\"3\\"><ins data-operation-index=\\"3\\">This is a new paragraph.</ins></p>
<div class=\\"notice notice-info\\">
<p>This is an info block</p>
</div>
<p><span class=\\"placeholder\\">This is a placeholder</span></p>
<p><span class=\\"highlight\\">this is a highlight</span></p>
<ul>
<li><span class=\\"checkbox\\">[<del data-operation-index=\\"5\\"> ]</del><ins data-operation-index=\\"5\\">x]</ins></span>checklist item 1</li><li><span class=\\"checkbox\\" data-diff-node=\\"ins\\" data-operation-index=\\"7\\"><ins data-operation-index=\\"7\\">[x]</ins></span><ins data-operation-index=\\"7\\">checklist item 2</ins></li>
<li><span class=\\"checkbox\\">[ ]</span>checklist item <del data-operation-index=\\"9\\">2</del><ins data-operation-index=\\"9\\">3</ins></li><li><span class=\\"checkbox\\" data-diff-node=\\"ins\\" data-operation-index=\\"9\\"><ins data-operation-index=\\"9\\">[ ]</ins></span><ins data-operation-index=\\"9\\">checklist item 4</ins></li>
<li><span class=\\"checkbox\\">[x]</span>checklist item <del data-operation-index=\\"11\\">3</del><ins data-operation-index=\\"11\\">5</ins></li>"
`;
